### ISSUE TITLE
Broken permalinks when host not set

### ISSUE THREAD

**Issue by josharian** [spf13/hugo#2987]
I'm setting up a hugo site for the first time, and the permalinks were wrong. I tracked down the error, but am not sure what the right fix is.

The problem occurs when a URL without a host is passed to URLPrep (helpers/url.go). The url gets passed to `purell.NormalizeURLString`, which parses out the host and passes it to `idna.ToASCII`. `idna.ToASCII` rejects the empty string, with error `idna: invalid label ""`. On non-nil error, URLPrep returns its original input, which includes the trailing `.html`, so the permalink is to the ugly url, not the pretty one. The rest of hugo doesn't know this, so the link is broken.

I see several possible fixes. Possibly multiple or all of them should be done.

* In URLPrep, instead of returning `in` on error, return `x`. Then the prettification gets kept. Also, print more context in the error message (the value of `x`, `in`, and `err`).
* In URLPrep, detect missing host and log an error message explaining how/where to set it. (I still don't know how. The docs are voluminous. I just want to get my site working.)
* In purell.NormalizeURLString, detect missing host and skip calling idna.ToASCII.
* In idna.ToASCII, when passed "", return "".

Input requested on the correct direction here. Thanks.

**Comment by bep** [comment]
That is one detailed issue report.

I think, however, it is better get a clear definition of the problem and a way to reproduce it before we talk about solving it, i.e: 

* What do you mean by "host is not set", something ala `http://:8080/mysite`?
* And If so, I assume this comes from the `baseURL` setting. Why do you need a hostless setup? 
* Is your site source located online for others to look at?

To recap the `baseURL` logic:

* The value set in `config.toml` (or similar) is the production value, i.e. the one used when you just run `hugo`
* When you run `hugo server` that value is overridden to `http://localhost:1313/` (and the server bound to `127.0.0.1`, which can overridden by `hugo server --baseURL=http://myhost:1313/ --bind=myhost``

**Comment by josharian** [comment]
Here's the top of my `config.toml`. It is not public yet, because I am still in the mess-around-with-it how-does-it-work stage.

```toml
languageCode = "en-us"
title = "Don't Panic"
baseURL = "https://commaok.xyz/"
theme = "hugo-cactus-theme"
```

I've done ~0 by way of configuration or work. Here are the hugo commands I've run:

```bash
$ go get github.com/spf13/hugo
$ cd $GOPATH/src/github.com/spf13/hugo/
$ hugo new site .
$ git clone https://github.com/digitalcraftsman/hugo-cactus-theme.git
$ hugo new post/new-entry.md
$ hugo server --buildDrafts
```

When I ran `hugo server`, I got a slew of warnings like:

```
ERROR 2017/01/24 22:07:14 Failed to normalize URL string (idna: invalid label ""). Returning in = "/", x = /
ERROR 2017/01/24 22:07:14 Failed to normalize URL string (idna: invalid label ""). Returning in = "/", x = /
ERROR 2017/01/24 22:07:14 Failed to normalize URL string (idna: invalid label ""). Returning in = "post", x = post
ERROR 2017/01/24 22:07:14 Failed to normalize URL string (idna: invalid label ""). Returning in = "post", x = post
ERROR 2017/01/24 22:07:14 Failed to normalize URL string (idna: invalid label ""). Returning in = "post/new-entry.html", x = post/compile-time-assertions
```

I then loaded localhost:1313 in the browser, clicked the link to my `new-entry` blog entry, and got a 404. The url bar showed a url ending in `.html`. Removing the trailing `.html` in the url bar loaded the correct page.

> What do you mean by "host is not set", something ala http://:8080/mysite?

Sorry, I mean that the value passed to `URLPrep` has no host--it is something like `"post/new-entry.html"`. I have no idea where that value came from.

> And If so, I assume this comes from the baseURL setting. Why do you need a hostless setup?

I don't need a hostless setup (although I do think that everything *should* work with just relative URLs, in principle).

I have no idea what I'm doing--I came new to hugo, am trying to get things working, got a bunch of errors, confirmed that I had `baseURL` set, was overwhelmed by the docs, and went looking at the source to make sense of what was actually happening. And what I saw looked suspicious to me, thus the detailed bug report.

**Comment by bep** [comment]
Those errors seem familiar, see: 

https://discuss.gohugo.io/t/started-getting-failed-to-normalize-url-string-returning-in/5034

I notice you install via `go get`. We fill eventually track down that dependency (if it doesn't fix itself), but until then you will either have to:

1. Use a prebuild binary
2. Build with `make` (we use `govendor`, but does not check in the deps source. You will need `govendor` and then maybe do a `govendor sync` or something.

There are also some issues reported with the latest master that may or may not make you want to use a release version.

**Comment by josharian** [comment]
OK, the problem was https://go-review.googlesource.com/c/34770/ which caused https://github.com/golang/go/issues/18567 which notes that one of the problem is that x/net/idna newly rejects the empty string, which caused the problems I saw above. It has now been reverted, in https://github.com/golang/net/commit/f2499483f923065a842d38eb4c7f1927e6fc6e6d. So anyone else who encounters this issue can simply update their x/net repo.

I do think you might want to consider returning the partially cleaned url in case of error in URLPrep anyway.

I looked at the installation docs for instructions for installing from source (both readme and the giant docs site), and saw nothing, just prebuilt downloads, so I assumed go get was fine. You might want to add a line or three.
